core/cap-0046.md

```diff
@@ -451,35 +514,96 @@ diff -ru '--exclude=*.h' '--exclude=.git*' '--exclude=*.md' src/protocol-curr/xd
  };
  
  /* CreateAccount
-@@ -465,6 +474,27 @@
+@@ -465,6 +474,88 @@
      int64 minAmountB; // minimum amount of second asset to withdraw
  };
  
-+enum HostFunction
++enum HostFunctionType
++{
++    HOST_FUNCTION_TYPE_INVOKE_CONTRACT = 0,
++    HOST_FUNCTION_TYPE_CREATE_CONTRACT = 1,
++    HOST_FUNCTION_TYPE_INSTALL_CONTRACT_CODE = 2
++};
++
++enum ContractIDType
++{
++    CONTRACT_ID_FROM_PUBLIC_KEY = 0,
++    CONTRACT_ID_FROM_ASSET = 1
++};
++ 
++enum ContractIDPublicKeyType
++{
++    CONTRACT_ID_PUBLIC_KEY_SOURCE_ACCOUNT = 0,
++    CONTRACT_ID_PUBLIC_KEY_ED25519 = 1
++};
++
++struct InstallContractCodeArgs
++{
++    opaque code<SCVAL_LIMIT>;
++};
++
++enum CreateContractSourceType {
++    CONTRACT_SOURCE_REF = 0,
++    CONTRACT_SOURCE_INSTALLED = 1
++};
++
++union CreateContractSource switch (CreateContractSourceType type)
 +{
-+    HOST_FN_INVOKE_CONTRACT = 0,
-+    HOST_FN_CREATE_CONTRACT_WITH_ED25519 = 1,
-+    HOST_FN_CREATE_CONTRACT_WITH_SOURCE_ACCOUNT = 2,
-+    HOST_FN_CREATE_TOKEN_CONTRACT_WITH_SOURCE_ACCOUNT = 3,
-+    HOST_FN_CREATE_TOKEN_CONTRACT_WITH_ASSET = 4
++case CONTRACT_SOURCE_REF:
++    SCContractCode codeRef;
++case CONTRACT_SOURCE_INSTALLED:
++    InstallContractCodeArgs installContractCodeArgs;
++};
++
++struct CreateContractArgs
++{
++    union switch (ContractIDType type)
++    {
++    case CONTRACT_ID_FROM_PUBLIC_KEY:
++        struct 
++        {
++            union switch (ContractIDPublicKeyType type)
++            {
++            case CONTRACT_ID_PUBLIC_KEY_SOURCE_ACCOUNT:
++                void;
++            case CONTRACT_ID_PUBLIC_KEY_ED25519:
++                struct
++                {
++                    uint256 key;
++                    Signature signature;
++                } ed25519KeyWithSignature;
++            } keySource;
++            uint256 salt;
++        } publicKey;
```

Comment:
Fyi I have this lurking memory that some of the SDKs don't do well with many nested layers of anonymous types, and we might need to make each layer in this XDR type a named type. We can see how it plays out first, but definitely remember experiencing some pain with structures defined anonymously with more than 2 levels deep in the past.